### PR TITLE
[BugFix] fix column overflow when handle too large partial update

### PR DESCRIPTION
**Pull request by luohaha**
## Why I'm doing:
In current implementation, when handle partial column update, we will try to build the column to be updated with segment granularity, it will lead to overflow. E.g.
1. If we have a `ARRAY` column to be update, In the beginning, because the data in the table does not contain arrays yet, we will slice the segment file with a size of 1GB, so a segment file may contain a large number of rows. We can assume that there are 500w rows in a tablet.
2. And in a `ArraryColumn` struct, we store offset of array using uint32_t , that means we can only have 4,294,967,295 items in a `ArraryColumn`.
3. When items in one array is larger than 900, which is large than 4,294,967,295, so overflow happens.

## What I'm doing:
Processing updates involving large amounts of data in batches, each batch will be limit by `partial_update_memory_limit_per_worker`.

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/storage/rowset_column_update_state.cpp [hunk 1]
```diff
@@ -300,11 +300,25 @@ Status RowsetColumnUpdateState::_finalize_partial_update_state(Tablet* tablet, R
...
 
-static StatusOr<ChunkPtr> read_from_source_segment(Rowset* rowset, const Schema& schema, Tablet* tablet,
-                                                   OlapReaderStatistics* stats, int64_t version,
-                                                   RowsetSegmentId rowset_seg_id, const std::string& path) {
+static int64_t calc_upt_memory_usage_per_row_column(Rowset* rowset) {
+    const auto& txn_meta = rowset->rowset_meta()->get_meta_pb_without_schema().txn_meta();
+    const int64_t total_update_col_cnt = txn_meta.partial_update_column_ids_size();
+    // `num_rows_upt` and `total_update_row_size` could be zero when upgrade from old version,
+    // then we will return zero and no limit.
+    if ((rowset->num_rows_upt() * total_update_col_cnt) <= 0) return 0;
```
Comment (chaoyli) [inline be/src/storage/rowset_column_update_state.cpp:308]: Whether `rowset->num_rows_upt() * total_update_col_cnt` will overflow?

Comment (luohaha) [reply be/src/storage/rowset_column_update_state.cpp:308]: No, they are both `int64_t` length

**Review by decster — DISMISSED** [review]

**Review by chaoyli — DISMISSED** [review]

**Review by wyb — COMMENTED** [review]
cloud native table also need this optimization?

File: be/src/storage/rowset/horizontal_update_rowset_writer.cpp [hunk 2]
```diff
@@ -58,7 +58,7 @@ Status HorizontalUpdateRowsetWriter::add_chunk(const Chunk& chunk) {
         RETURN_IF_ERROR(_update_file_writer->finalize(&segment_size, &index_size, &footer_position));
         {
             std::lock_guard<std::mutex> l(_lock);
-            _num_rows_upt += _update_file_writer->num_rows_written();
+            _num_rows_upt += chunk.num_rows();
```
[the remark this answers is not in the document]
Comment (luohaha) [reply be/src/storage/rowset/horizontal_update_rowset_writer.cpp]: done

**Review by wyb — APPROVED** [review]

**Review by decster — APPROVED** [review]

**Review by kevincai — APPROVED** [review]

**Review by dirtysalt — APPROVED** [review]

**Pull request merged**